util/chunk/row_container.go

```diff
@@ -0,0 +1,227 @@
+func (c *RowContainer) NumRowsOfChunk(chkID int) int {
+	if c.AlreadySpilled() {
+		return c.recordsInDisk.NumRowsOfChunk(chkID)
+	}
+	return c.records.NumRowsOfChunk(chkID)
+}
+
+// NumChunks returns the number of chunks in the container.
+func (c *RowContainer) NumChunks() int {
+	if c.AlreadySpilled() {
+		return c.recordsInDisk.NumChunks()
+	}
+	return c.records.NumChunks()
+}
+
+// Add appends a chunk into the RowContainer
+func (c *RowContainer) Add(chk *Chunk) (err error) {
+	if c.AlreadySpilled() {
+		err = c.recordsInDisk.Add(chk)
+	} else {
+		c.records.Add(chk)
+		if atomic.LoadUint32(&c.exceeded) != 0 {
+			err = c.spillToDisk()
+			if err != nil {
+				return err
+			}
+			c.records.Clear() // GC chunks in memory
+			atomic.StoreUint32(&c.spilled, 1)
+		}
+	}
+	return
+}
+
+// AppendRow appends one row into RowContainer, it is not thread safe
+func (c *RowContainer) AppendRow(row Row) (ptr RowPtr, err error) {
+	if c.AlreadySpilled() {
+		ptr, err = c.recordsInDisk.AppendRow(row)
+	} else {
+		ptr = c.records.AppendRow(row)
+		if atomic.LoadUint32(&c.exceeded) != 0 {
+			err = c.spillToDisk()
+			if err != nil {
+				return
+			}
+			c.records.Clear()
+			atomic.StoreUint32(&c.spilled, 1)
+		}
+	}
+	return
+}
+
+// GetChunk returns chkIdx th chunk of in memory records
+func (c *RowContainer) GetChunk(chkIdx int) *Chunk {
+	return c.records.GetChunk(chkIdx)
+}
+
+// GetRow returns the row the ptr pointed to
+func (c *RowContainer) GetRow(ptr RowPtr) (Row, error) {
+	if c.AlreadySpilled() {
+		return c.recordsInDisk.GetRow(ptr)
+	}
+	return c.records.GetRow(ptr), nil
+}
+
+// GetMemTracker returns the memory tracker in records, panics if the RowContainer has already spilled
+func (c *RowContainer) GetMemTracker() *memory.Tracker {
+	return c.records.GetMemTracker()
+}
+
+// GetDiskTracker returns the underlying disk usage tracker in recordsInDisk.
+func (c *RowContainer) GetDiskTracker() *disk.Tracker {
+	return c.recordsInDisk.GetDiskTracker()
+}
+
+// Close close the RowContainer
+func (c *RowContainer) Close() (err error) {
+	if c.AlreadySpilled() {
+		err = c.recordsInDisk.Close()
+		c.recordsInDisk = nil
+	}
+	return
+}
+
+// ActionSpill returns a memory.ActionOnExceed for spilling over to disk.
+func (c *RowContainer) ActionSpill() memory.ActionOnExceed {
+	return &spillDiskAction{c: c}
+}
+
+// spillDiskAction implements memory.ActionOnExceed for chunk.List. If
+// the memory quota of a query is exceeded, spillDiskAction.Action is
+// triggered.
+type spillDiskAction struct {
+	once           sync.Once
+	c              *RowContainer
+	fallbackAction memory.ActionOnExceed
+}
+
+// Action sends a signal to trigger spillToDisk method of hashRowContainer
+// and if it is already triggered before, call its fallbackAction.
+func (a *spillDiskAction) Action(t *memory.Tracker) {
+	if a.c.AlreadySpilledSafe() {
+		if a.fallbackAction != nil {
+			a.fallbackAction.Action(t)
+		}
+	}
+	a.once.Do(func() {
```

Comment:
We should make sure that the action can be triggered again after `RowContainer.Reset` is called.
And please add a unit test for this situation.